utils/render_utils.py

```diff
@@ -0,0 +1,289 @@
+  # Path height sits at z=0 (in middle of zero-mean capture pattern).
+  offset = np.array([center[0], center[1], 0])
+
+  # Calculate scaling for ellipse axes based on input camera positions.
+  sc = np.percentile(np.abs(poses[:, :3, 3] - offset), 90, axis=0)
+  # Use ellipse that is symmetric about the focal point in xy.
+  low = -sc + offset
+  high = sc + offset
+  # Optional height variation need not be symmetric
+  z_low = np.percentile((poses[:, :3, 3]), 10, axis=0)
+  z_high = np.percentile((poses[:, :3, 3]), 90, axis=0)
+
+  def get_positions(theta):
+    # Interpolate between bounds with trig functions to get ellipse in x-y.
+    # Optionally also interpolate in z to change camera height along path.
+    return np.stack([
+        low[0] + (high - low)[0] * (np.cos(theta) * .5 + .5),
+        low[1] + (high - low)[1] * (np.sin(theta) * .5 + .5),
+        z_variation * (z_low[2] + (z_high - z_low)[2] *
+                       (np.cos(theta + 2 * np.pi * z_phase) * .5 + .5)),
+    ], -1)
+
+  theta = np.linspace(0, 2. * np.pi, n_frames + 1, endpoint=True)
+  positions = get_positions(theta)
+
+  #if const_speed:
+
+  # # Resample theta angles so that the velocity is closer to constant.
+  # lengths = np.linalg.norm(positions[1:] - positions[:-1], axis=-1)
+  # theta = stepfun.sample(None, theta, np.log(lengths), n_frames + 1)
+  # positions = get_positions(theta)
+
+  # Throw away duplicated last position.
+  positions = positions[:-1]
+
+  # Set path's up vector to axis closest to average of input pose up vectors.
+  avg_up = poses[:, :3, 1].mean(0)
+  avg_up = avg_up / np.linalg.norm(avg_up)
+  ind_up = np.argmax(np.abs(avg_up))
+  up = np.eye(3)[ind_up] * np.sign(avg_up[ind_up])
+
+  return np.stack([viewmatrix(p - center, up, p) for p in positions])
+
+
+def generate_path(viewpoint_cameras, n_frames=480):
+  viewpoint_cameras = [viewpoint_cameras[i] for i in range(0, len(viewpoint_cameras), 300)]
+  
+  c2ws = np.array([np.linalg.inv(np.asarray((cam[1].world_view_transform.T).cpu().numpy())) for cam in viewpoint_cameras])
+  pose = c2ws[:,:3,:] @ np.diag([1, -1, -1, 1])
+  pose_recenter, colmap_to_world_transform = transform_poses_pca(pose)
+
+  # generate new poses
+  new_poses = generate_ellipse_path(poses=pose_recenter, n_frames=n_frames)
+  # warp back to orignal scale
+  new_poses = np.linalg.inv(colmap_to_world_transform) @ pad_poses(new_poses)
+
+  traj = []
+  for i, c2w in enumerate(new_poses):
+      c2w = c2w @ np.diag([1, -1, -1, 1])
+      cam = copy.deepcopy(viewpoint_cameras[0][1])
+      cam.image_height = int(cam.image_height / 2) * 2
+      cam.image_width = int(cam.image_width / 2) * 2
+      cam.world_view_transform = torch.from_numpy(np.linalg.inv(c2w).T).float().cuda()
+      cam.full_proj_transform = (cam.world_view_transform.unsqueeze(0).bmm(cam.projection_matrix.cuda().unsqueeze(0))).squeeze(0)
+      cam.camera_center = cam.world_view_transform.inverse()[3, :3]
+      cam.timestamp =10.0/n_frames*i
```

Comment:
Here the max timestamp value is hardcoded. This breaks other datasets which have a different scale(e.g dnerf with 0.0->1.0). I think you should be able to pass down the `time_duration` value from the config .yaml file

Reply:
Thank you for your review!
You’re absolutely right – I overlooked this point. I will make the necessary changes to pass the `time_duration` value from the config.yaml file instead of hardcoding the max timestamp.

Let me know if you have any further suggestions, and thanks again for your helpful feedback!